### ISSUE TITLE
Introduce a `requireModule("OpenSSL")` style API avoid pyflakes error.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7014]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)| @adiroiban reported|
|-|-|
|Trac ID|trac#7014|
|Type|enhancement|
|Created|2014-03-03 08:25:37Z|
|Branch|https://github.com/twisted/twisted/tree/require-module-api-7014|

This is created based on feedback from [#7002](https://github.com/twisted/twisted/issues/7002)

   1. twisted/conch/test/test_agent.py - I dislike adding useless code to
 placate stupid tools.  There are a number of possible better fixes here
 (perhaps any and all of which are better deferred to a follow-up ticket):
     1. Introduce a `requireModule("OpenSSL")` style API so we don't have
 to write code that pyflakes is sure is useless.  This is both more
 expressive to pyflakes and to future maintainers of the code.

Look for ```# FIXME: #7014``` comments after this is done. 

Attachments:

* [7014-1.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/390/3905318ce00b80713cee6592a2c4d52573d58f66/9bdf72b95745d04c08b363298c9b4aa830e92368.diff) (3746 bytes) - added by adiroiban on 2014-03-10 09:28:11Z - 
* [7014-2.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/390/3905318ce00b80713cee6592a2c4d52573d58f66/27c2b15dcfd60cad03295a85bdefbdd05f44ac36.diff) (4018 bytes) - added by adiroiban on 2014-03-10 09:37:12Z - 
* [7014-3.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/390/3905318ce00b80713cee6592a2c4d52573d58f66/8367b598a5ca1c5a96f6f2119218094c574870b2.diff) (927 bytes) - added by adiroiban on 2014-03-16 16:41:43Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__7014 7014
type__enhancement enhancement
reporter__adiroiban adiroiban
priority__normal normal
milestone__ 
branch__branches_require_module_api_7014 branches/require-module-api-7014
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1393835137000000 1393835137000000
changetime__1395030485000000 1395030485000000
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)|@adiroiban set owner to @adiroiban|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)|@adiroiban removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_4"></a>|@adiroiban commented|
|-|-|

Please check latest diff.

I have implemented requireModule(name, default=None) API. It only returns default value when ImportError is raised. ValueError or other errors are not handled.

I added tests for the new method, but I don't know where to write tests for ```__all__``` member.

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl commented|
|-|-|

(In [41909]) Branching to require-module-api-7014.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl set owner to @adiroiban|
|-|-|

Hi adiroiban, thanks for this patch!

I ported it to use the new reflect.py location (_reflectpy3.py is now gone), and got it to run on the buildbots. There's currently a failing test - "twisted.trial.unittest.FailTest: ValueError not raised (None returned)" in twisted.test.test_reflect.test_requireModuleOtherError, on Python 3.

I'd recommend basing any further patches off `branches/require-module-api-7014`.

Please fix this failing test and resubmit for review.

-hawkowl

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_7"></a>|@adiroiban removed owner|
|-|-|

I have attached a diff on top of dedicated branch.

I have removed the ValueError tests since that behavior was only present in Python 2.7 ... on 3.3 it always get ImportError

Thanks for the review!

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_8"></a>|@hawkowl set owner to @hawkowl|
|-|-|

Hi,

Thanks - I'm not so sure about removing the test completely, but I think that passing weird things to `__import__` is in "undefined behaviour" territory anyway (who knows how it acts on IronPython??), so I think it's fine.

Looks good to me, I'll merge.

-hawkowl

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl set status to `closed`|
|-|-|

(In [41915]) Merge require-module-api-7014: A requireModule() style API

Author: adiroiban
Reviewer: hawkowl
Fixes: [#7014](https://github.com/twisted/twisted/issues/7014)

Introduces an API to allow conditionally importing optional modules.
